PowerShellTools/LanguageService/SmartIndent.cs

```
@@ -0,0 +1,167 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Editor.OptionsExtensionMethods;
+using Microsoft.VisualStudio.TextManager.Interop;
+using PowerShellTools.Classification;
+
+namespace PowerShellTools.LanguageService
+{
+    internal sealed class SmartIndent : ISmartIndent
+    {
+        private ITextView _textView;
+        private PowerShellLanguageInfo _info;
+
+	/// <summary>
+	/// Constructor.
+	/// </summary>
+	/// <param name="info">Powershell language service.</param>
+	/// <param name="textView">Current active TextView.</param>
+        public SmartIndent(PowerShellLanguageInfo info, ITextView textView)
+        {
+            _info = info;
+            _textView = textView;
+        }
+
+	/// <summary>
+	/// Implementation of the interface.
+	/// </summary>
+	/// <param name="line">The current line after Enter.</param>
+	/// <returns>Desired indentation size.</returns>
+        public int? GetDesiredIndentation(ITextSnapshotLine line)
+        {
+
+	    // User GetIndentSize() instead of GetTabSize() due to the fact VS always uses Indent Size as a TAB size
+	    int tabSize = _textView.Options.GetIndentSize();
+
+            try
+            {
+                switch (_info.LangPrefs.IndentMode)
+                {
+                    case vsIndentStyle.vsIndentStyleNone:
+                        return null;
+
+                    case vsIndentStyle.vsIndentStyleDefault:
+                        return GetDefaultIndentationImp(line, tabSize);
+
+                    case vsIndentStyle.vsIndentStyleSmart:
+                        return GetSmartIndentationImp(line, tabSize);
+                }
+                return null;
+            }
+            catch

```

We should report an error, instead of hiding it...